### PR TITLE
Optional compile flag to exclude init mode

### PR DESCRIPTION
**Pull request by mark-petersen**
Currently, when MPAS is run in E3SM, it looks for all the init mode namelists in the namelist file. When they are not found, the MPAS framework writes lengthly warnings to the ocean log file that are not relevant to E3SM, and are confusing to the E3SM user.

This PR adds the option to exclude init mode in both the Make and cmake systems. The default Makefile build remains the same as before, but init mode may be excluded with
```
make $COMPILER CORE=ocean EXCLUDE_INIT_MODE=true
```

The cmake file is currently only used in E3SM, and was altered to not include init mode.

This may also speed up E3SM compile time because the init mode files will no longer be compiled in E3SM.

### CODE REVIEW

**Comment by mark-petersen** [comment]
<details>
<summary> Currently the file `ocn.log.*` in the E3SM run directory starts out with this: </summary> <p>

```
 *** Encountered an issue while attempting to read namelist record lock_exchange
     The following values will be used for variables in this record:

         config_lock_exchange_vert_levels = 20
         config_lock_exchange_bottom_depth = 20.0000000000000
         config_lock_exchange_cold_temperature = 5.00000000000000
         config_lock_exchange_warm_temperature = 30.0000000000000
         config_lock_exchange_direction = y
         config_lock_exchange_salinity = 35.0000000000000
         config_lock_exchange_layer_type = z-level
         config_lock_exchange_isopycnal_min_thickness = 0.999999977648258E-02

 *** Encountered an issue while attempting to read namelist record internal_waves
     The following values will be used for variables in this record:

         config_internal_waves_vert_levels = 20
         config_internal_waves_use_distances = F
         config_internal_waves_surface_temperature = 20.1000003814697
         config_internal_waves_bottom_temperature = 10.1000003814697
         config_internal_waves_temperature_difference = 2.00000000000000
         config_internal_waves_amplitude_width_frac = 0.330000013113022
         config_internal_waves_amplitude_width_dist = 50000.0000000000
         config_internal_waves_bottom_depth = 500.000000000000
         config_internal_waves_salinity = 35.0000000000000
         config_internal_waves_layer_type = z-level
         config_internal_waves_isopycnal_displacement = 125.000000000000

 *** Encountered an issue while attempting to read namelist record overflow
     The following values will be used for variables in this record:

         config_overflow_vert_levels = 100
         config_overflow_use_distances = F
         config_overflow_bottom_depth = 2000.00000000000
         config_overflow_ridge_depth = 500.000000000000
         config_overflow_plug_temperature = 10.0000000000000
         config_overflow_domain_temperature = 20.0000000000000
         config_overflow_salinity = 35.0000000000000
         config_overflow_plug_width_frac = 0.100000001490116
         config_overflow_slope_center_frac = 0.200000002980232
         config_overflow_slope_width_frac = 0.500000007450581E-01
         config_overflow_plug_width_dist = 20000.0000000000
         config_overflow_slope_center_dist = 40000.0000000000
         config_overflow_slope_width_dist = 7000.00000000000
         config_overflow_layer_type = z-level
         config_overflow_isopycnal_min_thickness = 0.999999977648258E-02

 *** Encountered an issue while attempting to read namelist record dam_break
     The following values will be used for variables in this record:

         config_dam_break_vert_levels = 1
         config_dam_break_eta0 = 0.600000023841858
         config_dam_break_dc = 0.399999991059303E-01
         config_dam_break_R0 = 24.2000007629395
         config_dam_break_Xl = 1.00000000000000
         config_dam_break_Yl = 2.00000000000000
         config_dam_break_Inlet = 0.400000005960464

 *** Encountered an issue while attempting to read namelist record global_ocean
     The following values will be used for variables in this record:

         config_global_ocean_minimum_depth = 15.0000000000000
         config_global_ocean_depth_file = vertical_grid.nc
         config_global_ocean_depth_dimname = nVertLevels
         config_global_ocean_depth_varname = refMidDepth
         config_global_ocean_depth_conversion_factor = 1.00000000000000
         config_global_ocean_temperature_file = none
         config_global_ocean_salinity_file = none
         config_global_ocean_tracer_nlat_dimname = none
         config_global_ocean_tracer_nlon_dimname = none
         config_global_ocean_tracer_ndepth_dimname = none
```
</details>
and on and on, to line 550. I have seen this baffle new E3SM users. 

I tested this PR in stand-alone and in E3SM with:
```
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_gnu
```
and it removes the init mode portions of this, as expected.

**Comment by mark-petersen** [comment]
Rebased. Tested in both stand-alone ocean and E3SM, with both intel and gnu. 

When compiled with `MODEL_FORMULATION=-DEXCLUDE_INIT_MODE`, init mode is excluded, as expected. When compiled without that flag, behavior is identical to before for nightly regression suite.

In E3SM, passes
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_gnu
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
```
and error messages about init mode flags in ocn.log file no longer appear.

**Comment by matthewhoffman** [comment]
@mark-petersen , as we discussed on the call, you can alternatively implement this as a custom variable on the make line, because all variables defined there get passed down to the core makefiles.  Here is the example in the landice core of how we compile with Albany by doing `make gfortran CORE=landice ALBANY=true`.  This gets handled here in the standalone make system:
https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/src/core_landice/build_options.mk#L21-L24
and here in cmake:
https://github.com/MPAS-Dev/MPAS-Model/blob/landice/develop/src/core_landice/landice.cmake#L16-L18

**Comment by mark-petersen** [comment]
Thanks @matthewhoffman for the suggestion. I changed it to use `EXCLUDE_INIT_MODE=true`.

@xylar and @mattdturner this is ready to review.

File: src/core_ocean/Registry.xml [hunk 1]
```diff
@@ -1302,6 +1302,12 @@
 					possible_values="'linear_x', 'linear_y', 'linear_arb_rot', 'power_x', 'power_y', 'power_arb_rot', 'sin_arb_rot', 'sph_solid_body', 'sph_Williamson', 'sph_uCosCos', 'sph_vCosCos', 'sph_ELonLon_CosCos', 'sph_ELatLat_CosCos', 'sph_ELonLat_CosCos'"
 		/>
 	</nml_record>
+	<nml_record name="init_mode_vert_levels" mode="init">
+		<nml_option name="config_vert_levels" type="integer" default_value="-1" units="unitless"
```
Comment (mark-petersen) [inline src/core_ocean/Registry.xml:1306]: I had to leave this flag in the main Registry file because it is used to fill in a dimension.

**Review by xylar — CHANGES_REQUESTED** [review]
With the fix mentioned above and with gnu compilers on my Ubuntu laptop, I was able to:
* compile without `EXCLUDE_INIT_MODE=true` and run the legacy COMPASS nightly tests, which passed bit-for-bit compared with a baseline from the current `ocean/develop`.
* compile with `EXCLUDE_INIT_MODE=true` and run the QU240 test cases from the legacy COMPASS nightly regression suite other than `init` (which obviously can't be set up without init mode).  The tests were also bit-for-bit

**Comment by mark-petersen** [comment]
@xylar thanks, I added that. I also had to add a few lines to the CPP registry parsing. Could you re-test?

File: src/core_ocean/Makefile [hunk 2]
```diff
@@ -1,35 +1,46 @@
...
+endif
 	(cd driver; $(MAKE) FCINCLUDES="$(FCINCLUDES) $(OCEAN_SHARED_INCLUDES)" all )
 	if [ -e libdycore.a ]; then \
 		($(RM) libdycore.a) \
 	fi
 	ar -ru libdycore.a `find . -type f -name "*.o"`
 
 core_reg:
-	$(CPP) $(CPPFLAGS) $(CPPINCLUDES) Registry.xml > Registry_processed.xml
+	$(CPP) $(CPPFLAGS) $(CPPLOCALFLAGS) $(CPPINCLUDES) Registry.xml > Registry_processed.xml)
```
Comment (xylar) [inline src/core_ocean/Makefile:27]:
```suggestion
	$(CPP) $(CPPFLAGS) $(CPPLOCALFLAGS) $(CPPINCLUDES) Registry.xml > Registry_processed.xml
```
@mark-petersen, you added an extraneous parenthesis here.

File: src/core_ocean/Registry.xml (hunk 1 above)
Comment (xylar) [reply src/core_ocean/Registry.xml:1306]: I thought about this and looked through the code.  I can't come up with a cleaner solution than this either.

**Comment by xylar** [comment]
Retesting now...

**Comment by mattdturner** [comment]
On Compy:
```
PASS:
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi.C
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi BASELINE /compyfs/e3sm_baselines/pgi/master: 
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
```

**Comment by mark-petersen** [comment]
Passes:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_gnu
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
PEM_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.cori-haswell_gnu
PET_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu
```
and `ocn.log*` file no longer shows init mode error messages, as expected.

**Review by xylar — APPROVED** [review]
I rebuilt with the latest version of this branch on my Ubuntu laptop with Gnu.  All legacy COMPASS  nightly tests passed without `EXCLUDE_INIT_MODE=true`.  All non-init QU240 tests passed with `EXCLUDE_INIT_MODE=true`.

**Review by mattdturner — APPROVED** [review]

**Pull request merged**